Mainnet/AddressMapper/AddressMapper/AddressMapper.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using Stratis.SmartContracts;
+
+public class AddressMapper : SmartContract
+{
+    public Address GetSecondaryAddress(Address primary) => PersistentState.GetAddress($"SecondaryAddress:{primary}");
+    private void SetSecondaryAddress(Address primary, Address secondary) => PersistentState.SetAddress($"SecondaryAddress:{primary}", secondary);
+
+    private MappingInfo GetMappingInfo(Address secondary) => PersistentState.GetStruct<MappingInfo>($"MappingInfo:{secondary}");
+    private void SetMappingInfo(Address secondary, MappingInfo value) => PersistentState.SetStruct($"MappingInfo:{secondary}", value);
+
+    private void ClearMappingInfo(Address secondary) => PersistentState.Clear($"MappingInfo:{secondary}");
+
+    public Address Owner
+    {
+        get => PersistentState.GetAddress(nameof(Owner));
+        private set => PersistentState.SetAddress(nameof(Owner), value);
+    }
+
+    public AddressMapper(ISmartContractState smartContractState, Address owner) : base(smartContractState)
+    {
+        this.Owner = owner;
+    }
+
+    public void MapAddress(Address secondary)
+    {
+        Assert(GetMappingInfo(secondary).Status == (int)Status.NoStatus);
+
+        SetMappingInfo(secondary, new MappingInfo { Primary = Message.Sender, Status = (int)Status.Pending });
+    }
+
+    public void Approve(Address secondary)
+    {
+        EnsureAdminOnly();
+        var mapping = GetMappingInfo(secondary);
+        Assert(mapping.Status == (int)Status.Pending, "Mapping is not in pending state.");
+
+        SetSecondaryAddress(mapping.Primary, secondary);
+        SetMappingInfo(secondary, new MappingInfo { Primary = mapping.Primary, Status = (int)Status.Approved });
+
+        Log(new AddressMappedLog { Primary = mapping.Primary, Secondary = secondary });
+    }
+
+    public void Reject(Address secondary)
+    {
+        EnsureAdminOnly();
+        var mapping = GetMappingInfo(secondary);
+        Assert(mapping.Status == (int)Status.Pending, "Mapping is not in pending state.");
+
+        ClearMappingInfo(secondary); // same address can be mapped again. 
+    }
+
+    public string GetStatus(Address secondary) => GetMappingInfo(secondary).Status.ToString();
+    public Address GetPrimaryAddress(Address secondary) => GetMappingInfo(secondary).Primary;
```

Comment:
This will map the secondary=>primary address even if the mapping has not been approved. This is inconsistent: we can't map primary=>secondary by calling `GetSecondaryAddress` until the mapping is approved.

Reply:
> I think `GetPrimaryAddress` needs to check for a valid approval.
> 
> The DoS vector may not be likely in practice but would be easy to accomplish. And it's avoidable with a design that allows for multiple mappings for a secondary address, leaving it up to the owner to decide which is real and which is not.

Currently `GetPrimaryAddress` return the result whether it is approved or pending. So client can get status by calling `GetStatus`. This is how Identity and STO Platform works. But method naming is confusing bit of. I think this can be design as you requested way as well. 

There was reason we have gone in this style of design but i think this is not necessary after seeing use case of both clients so i will update the code as you requested way.